langlib/lang.array/src/main/java/org/ballerinalang/langlib/array/Sort.java

```diff
@@ -18,73 +18,185 @@
+                        m.getTag() == TypeTags.STRING_TAG || m.getTag() == TypeTags.BOOLEAN_TAG ||
+                        m.getTag() == TypeTags.ARRAY_TAG).findFirst();
+        if (mainType.isPresent()) {
+            type = mainType.get();
+        }
+        return type;
+    }
+
+    private static void mergesort(Object[][] input, Object[][] aux, int lo, int hi, boolean isAscending, BType type) {
         if (hi <= lo) {
             return;
         }
 
         int mid = lo + (hi - lo) / 2;
 
-        mergesort(input, aux, lo, mid, strand, comparator);
-        mergesort(input, aux, mid + 1, hi, strand, comparator);
+        mergesort(input, aux, lo, mid, isAscending, type);
+        mergesort(input, aux, mid + 1, hi, isAscending, type);
 
-        merge(input, aux, lo, mid, hi, strand, comparator);
+        merge(input, aux, lo, mid, hi, isAscending, type);
     }
 
-    private static void merge(ArrayValue input, ArrayValue aux, int lo, int mid, int hi, Strand strand,
-                              FPValue<Object, Long> comparator) {
-        int elemTypeTag = input.getElementType().getTag();
-
-        for (int i = lo; i <= hi; i++) {
-            aux.add(i, input.get(i));
+    private static void merge(Object[][] input, Object[][] aux, int lo, int mid, int hi, boolean isAscending,
+                              BType type) {
+        if (hi + 1 - lo >= 0) {
+            System.arraycopy(input, lo, aux, lo, hi + 1 - lo);
         }
 
         for (int i = lo, j = mid + 1, k = lo; k <= hi; k++) {
+            int index;
             if (i > mid) {
-                input.add(k, aux.get(j++));
+                index = j++;
             } else if (j > hi) {
-                input.add(k, aux.get(i++));
-            } else if (comparator.call(new Object[]{strand, aux.get(j), true, aux.get(i), true}) < 0) {
-                input.add(k, aux.get(j++));
+                index = i++;
+            } else if (isAscending && sortFunc(aux[j][0], aux[i][0], type, true) < 0) {
+                index = j++;
+            } else if (!isAscending && sortFunc(aux[i][0], aux[j][0], type, false) < 0) {
+                index = j++;
             } else {
-                input.add(k, aux.get(i++));
+                index = i++;
             }
+            input[k] = aux[index];
         }
     }
 
+    private static int sortFunc(Object value1, Object value2, BType type, boolean isAscending) {
+        int c = -999;
+        // () should come last irrespective of the sort direction.
+        if (value1 == null) {
+            if (value2 == null) {
+                return 0;
+            }
+            if (isAscending) {
+                return 1;
+            }
+            return -1;
+        } else if (value2 == null) {
+            if (isAscending) {
+                return -1;
+            }
+            return 1;
+        } else if (type.getTag() == TypeTags.INT_TAG) {
```

Comment:
What about subtypes like `SIGNED32_INT_TAG`, `SIGNED16_INT_TAG`, etc.? Same for string and char too.

The util methods in TypeTags.java may be useful. org.ballerinalang.jvm.types.TypeTags#isIntegerTypeTag